packages/shorebird_cli/test/src/commands/doctor_command_test.dart

```diff
@@ -1,83 +1,91 @@
+    late ShorebirdVersionValidator shorebirdVersionValidator;
 
     setUp(() {
       logger = _MockLogger();
-      fetchCurrentVersionResult = _MockProcessResult();
-      fetchLatestVersionResult = _MockProcessResult();
+      progress = _MockProgress();
+
+      androidInternetPermissionValidator =
+          _MockAndroidInternetPermissionValidator();
+      shorebirdVersionValidator = _MockShorebirdVersionValidator();
 
       command = DoctorCommand(
         logger: logger,
-        runProcess: (
-          executable,
-          arguments, {
-          bool runInShell = false,
-          workingDirectory,
-        }) async {
-          if (executable == 'git') {
-            const revParseHead = ['rev-parse', '--verify', 'HEAD'];
-            if (arguments.every((arg) => revParseHead.contains(arg))) {
-              return fetchCurrentVersionResult;
-            }
-
-            const revParseUpstream = ['rev-parse', '--verify', '@{upstream}'];
-            if (arguments.every((arg) => revParseUpstream.contains(arg))) {
-              return fetchLatestVersionResult;
-            }
-          }
-          return _MockProcessResult();
-        },
+        validators: [
+          androidInternetPermissionValidator,
+          shorebirdVersionValidator,
+        ],
       );
 
-      when(
-        () => fetchCurrentVersionResult.exitCode,
-      ).thenReturn(ExitCode.success.code);
-      when(
-        () => fetchCurrentVersionResult.stdout,
-      ).thenReturn(currentShorebirdRevision);
-      when(
-        () => fetchLatestVersionResult.exitCode,
-      ).thenReturn(ExitCode.success.code);
-      when(
-        () => fetchLatestVersionResult.stdout,
-      ).thenReturn(currentShorebirdRevision);
+      when(() => logger.progress(any())).thenReturn(progress);
+
+      when(() => logger.info(any())).thenReturn(null);
+
+      when(() => androidInternetPermissionValidator.description)
+          .thenReturn('Android');
+      when(() => androidInternetPermissionValidator.validate())
+          .thenAnswer((_) async => []);
+
+      when(() => shorebirdVersionValidator.description)
+          .thenReturn('Shorebird Version');
+      when(() => shorebirdVersionValidator.validate())
+          .thenAnswer((_) async => []);
     });
 
     test('prints "no issues" when everything is OK', () async {
       await command.run();
+      for (final validator in command.validators) {
+        verify(validator.validate).called(1);
+      }
       verify(
         () => logger.info(captureAny(that: contains('No issues detected'))),
       ).called(1);
     });
 
-    test('prints that an upgrade is available', () async {
+    test('prints "x issues found" when warnings or errors found', () async {
       when(
-        () => fetchLatestVersionResult.stdout,
-      ).thenReturn(newerShorebirdRevision);
+        () => androidInternetPermissionValidator.validate(),
+      ).thenAnswer(
+        (_) async => [
+          const ValidationIssue(
```

Comment:
Minor but thoughts on adding a warning in here as well?

Reply:
Done.